[BUG] Web Features job failing - 2026-02-18

Culprit log line:

```
ERROR invalid status while batch writing status="rpc error: code = InvalidArgument desc = Row [text-wrap] in SystemManagedSavedSearches was already inserted in this transaction."
```

Logs: https://cloudlogging.app.goo.gl/kXVdXPzkQgzmmAUB7

--- a/lib/gcpspanner/saved_search_subscription.go
+++ b/lib/gcpspanner/saved_search_subscription.go
@@ -363,6 +363,34 @@ func (c *Client) ListSavedSearchSubscriptions(
 	return newEntityLister[savedSearchSubscriptionMapper](c).list(ctx, req)
 }
 
+func (c *Client) ListSubscriptionsBySavedSearchID(
+	ctx context.Context,
+	savedSearchID string,
+) ([]SavedSearchSubscription, error) {
+	stmt := spanner.NewStatement(`
+		SELECT ID, ChannelID, SavedSearchID, Triggers, Frequency, CreatedAt, UpdatedAt
+		FROM SavedSearchSubscriptions
+		WHERE SavedSearchID = @savedSearchID
+	`)
+	stmt.Params["savedSearchID"] = savedSearchID
+
+	iter := c.Single().Query(ctx, stmt)
+	defer iter.Stop()
+
+	var results []SavedSearchSubscription
+	err := iter.Do(func(row *spanner.Row) error {
+		var sub SavedSearchSubscription
+		if err := row.ToStruct(&sub); err != nil {
+			return err
+		}
+		results = append(results, sub)
+
+		return nil
+	})
+
+	return results, err
+}
+
 type spannerSubscriberDestination struct {
 	SubscriptionID string                  `spanner:"ID"`
 	UserID         string                  `spanner:"UserID"`

--- a/lib/gcpspanner/web_features.go
+++ b/lib/gcpspanner/web_features.go
@@ -340,7 +340,7 @@ func (m webFeatureSpannerMapper) moveLatestFeatureDeveloperSignals(
 func (m webFeatureSpannerMapper) moveSystemManagedSavedSearch(
 	ctx context.Context,
 	c *Client,
-	sourceID, targetKey string,
+	sourceID, targetID, targetKey string,
 	savedSearchMutations *[]*spanner.Mutation,
 	systemManagedSearchMutations *[]*spanner.Mutation,
 ) error {
@@ -354,6 +354,57 @@ func (m webFeatureSpannerMapper) moveSystemManagedSavedSearch(
 
 		return fmt.Errorf("unable to get system managed saved search: %w", err)
 	}
+
+	// Check if the target feature already has a system managed saved search.
+	// If it does, we don't want to overwrite it with the redirected one.
+	targetSystemManagedSearch, err := c.GetSystemManagedSavedSearchByFeatureID(ctx, targetID)
+	if err == nil {
+		// Target exists. Migrate subscriptions from the old saved search to the new one.
+		subscriptions, err := c.ListSubscriptionsBySavedSearchID(ctx, systemManagedSearch.SavedSearchID)
+		if err != nil {
+			return fmt.Errorf("failed to list subscriptions for migration: %w", err)
+		}
+
+		// Fetch existing subscriptions for the target saved search to check for duplicates.
+		targetSubscriptions, err := c.ListSubscriptionsBySavedSearchID(ctx, targetSystemManagedSearch.SavedSearchID)
+		if err != nil {
+			return fmt.Errorf("failed to list target subscriptions for deduplication: %w", err)
+		}
+		targetChannelIDs := make(map[string]bool)
+		for _, s := range targetSubscriptions {
+			targetChannelIDs[s.ChannelID] = true
+		}
+
+		for _, sub := range subscriptions {
+			if targetChannelIDs[sub.ChannelID] {
+				// User is already subscribed to the target feature. Skip migration (effectively deleting the old subscription).
+				continue
+			}
+
+			sub.SavedSearchID = targetSystemManagedSearch.SavedSearchID
+			sub.UpdatedAt = spanner.CommitTimestamp
+			// We update the subscription to point to the new saved search ID.
+			m, err := spanner.UpdateStruct(savedSearchSubscriptionTable, sub)
+			if err != nil {
+				return fmt.Errorf("failed to create update mutation for subscription migration: %w", err)
+			}
+			*savedSearchMutations = append(*savedSearchMutations, m)
+		}
+
+		// We do NOT delete the source system managed saved search association here.
+		// By leaving it, GetChildDeleteKeyMutations will find it later and delete it
+		// ALONG WITH the old SavedSearch (which is what we want, as we are done with it).
+		return nil
+	}
+	if !errors.Is(err, ErrQueryReturnedNoResults) {
+		return fmt.Errorf("unable to check for existing target system managed saved search: %w", err)
+	}
+
+	// We are repurposing the source saved search for the target.
+	// We MUST delete the old association here to prevent GetChildDeleteKeyMutations from
+	// finding it and deleting the SavedSearch row we are about to update.
+	deleteSubMutation := spanner.Delete(systemManagedSavedSearchesTable, spanner.Key{sourceID})
+	*systemManagedSearchMutations = append(*systemManagedSearchMutations, deleteSubMutation)
 
 	savedSearch, err := c.GetSavedSearch(ctx, systemManagedSearch.SavedSearchID)
 	if err != nil {
@@ -370,19 +421,18 @@ func (m webFeatureSpannerMapper) moveSystemManagedSavedSearch(
 	}
 	*savedSearchMutations = append(*savedSearchMutations, updateMutation)
 
-	// Now update the system-managed saved search association
-	deleteSubMutation := spanner.Delete(systemManagedSavedSearchesTable, spanner.Key{sourceID})
 	newSystemManagedSearch := SystemManagedSavedSearch{
-		FeatureID:     targetKey, // The target ID (feature key) is the new FeatureID
+		FeatureID:     targetID,
 		SavedSearchID: systemManagedSearch.SavedSearchID,
 		CreatedAt:     spanner.CommitTimestamp,
 		UpdatedAt:     spanner.CommitTimestamp,
 	}
-	insertSubMutation, err := spanner.InsertStruct(systemManagedSavedSearchesTable, &newSystemManagedSearch)
+	// Use InsertOrUpdate to handle potential race conditions or multiple redirects to the same target in the same batch.
+	insertSubMutation, err := spanner.InsertOrUpdateStruct(systemManagedSavedSearchesTable, &newSystemManagedSearch)
 	if err != nil {
 		return fmt.Errorf("unable to create insert mutation for system managed search: %w", err)
 	}
-	*systemManagedSearchMutations = append(*systemManagedSearchMutations, deleteSubMutation, insertSubMutation)
+	*systemManagedSearchMutations = append(*systemManagedSearchMutations, insertSubMutation)
 
 	return nil
 }
@@ -467,7 +517,7 @@ func (m webFeatureSpannerMapper) PreDeleteHook(
 
 		// Now move the SystemManagedSavedSearch
 		if err := m.moveSystemManagedSavedSearch(
-			ctx, c, sourceID, targetKey, &savedSearchMutations, &systemManagedSearchMutations); err != nil {
+			ctx, c, sourceID, targetID, targetKey, &savedSearchMutations, &systemManagedSearchMutations); err != nil {
 			slog.ErrorContext(ctx, "failed to move system managed saved search", "error", err, "sourceID", sourceID)
 
 			return nil, err

--- a/lib/gcpspanner/web_features_test.go
+++ b/lib/gcpspanner/web_features_test.go
@@ -758,3 +758,162 @@ func TestSyncWebFeatures_RedirectForMissingSource(t *testing.T) {
 		t.Errorf("features mismatch (-want +got):\n%s", diff)
 	}
 }
+
+func TestSyncWebFeatures_MultipleRedirectsToSameTarget(t *testing.T) {
+	ctx := context.Background()
+	restartDatabaseContainer(t)
+
+	// 1. Setup initial state with three features
+	initialState := []WebFeature{
+		{FeatureKey: "feature-a", Name: "Feature A", Description: "", DescriptionHTML: ""},
+		{FeatureKey: "feature-b", Name: "Feature B", Description: "", DescriptionHTML: ""},
+		{FeatureKey: "feature-c", Name: "Feature C", Description: "", DescriptionHTML: ""},
+	}
+	if err := spannerClient.SyncWebFeatures(ctx, initialState); err != nil {
+		t.Fatalf("Failed to set up initial state: %v", err)
+	}
+
+	// Ensure system managed saved searches exist
+	if err := spannerClient.SyncSystemManagedSavedQuery(ctx); err != nil {
+		t.Fatalf("Failed to sync system managed saved queries: %v", err)
+	}
+
+	// Get IDs for the features
+	pairs, err := spannerClient.FetchAllWebFeatureIDsAndKeys(ctx)
+	if err != nil {
+		t.Fatalf("Failed to fetch feature IDs and keys: %v", err)
+	}
+	featureKeyToID := make(map[string]string)
+	for _, pair := range pairs {
+		featureKeyToID[pair.FeatureKey] = pair.ID
+	}
+
+	// Verify initial saved searches exist
+	smsA, err := spannerClient.GetSystemManagedSavedSearchByFeatureID(ctx, featureKeyToID["feature-a"])
+	if err != nil {
+		t.Fatalf("Failed to get system managed saved search for feature-a: %v", err)
+	}
+	smsB, err := spannerClient.GetSystemManagedSavedSearchByFeatureID(ctx, featureKeyToID["feature-b"])
+	if err != nil {
+		t.Fatalf("Failed to get system managed saved search for feature-b: %v", err)
+	}
+	smsC, err := spannerClient.GetSystemManagedSavedSearchByFeatureID(ctx, featureKeyToID["feature-c"])
+	if err != nil {
+		t.Fatalf("Failed to get system managed saved search for feature-c: %v", err)
+	}
+
+	// --- Subscription Setup ---
+	userID := "test-user"
+	channelID, err := spannerClient.CreateNotificationChannel(ctx, CreateNotificationChannelRequest{
+		UserID: userID,
+		Name:   "Test Channel",
+		Type:   NotificationChannelTypeEmail,
+		EmailConfig: &EmailConfig{
+			Address:           "test@example.com",
+			IsVerified:        true,
+			VerificationToken: nil,
+		},
+	})
+	if err != nil {
+		t.Fatalf("Failed to create notification channel: %v", err)
+	}
+
+	subA, err := spannerClient.CreateSavedSearchSubscription(ctx, CreateSavedSearchSubscriptionRequest{
+		UserID:        userID,
+		ChannelID:     *channelID,
+		SavedSearchID: smsA.SavedSearchID,
+		Triggers:      []SubscriptionTrigger{SubscriptionTriggerBrowserImplementationAnyComplete},
+		Frequency:     SavedSearchSnapshotTypeWeekly,
+	})
+	if err != nil {
+		t.Fatalf("Failed to create subscription for feature-a: %v", err)
+	}
+
+	// Create a subscription for feature-c (target) to test deduplication.
+	// Since the user is already subscribed to the target, the subscription for feature-a should be dropped
+	// during migration.
+	subC, err := spannerClient.CreateSavedSearchSubscription(ctx, CreateSavedSearchSubscriptionRequest{
+		UserID:        userID,
+		ChannelID:     *channelID,
+		SavedSearchID: smsC.SavedSearchID,
+		Triggers:      []SubscriptionTrigger{SubscriptionTriggerBrowserImplementationAnyComplete},
+		Frequency:     SavedSearchSnapshotTypeWeekly,
+	})
+	if err != nil {
+		t.Fatalf("Failed to create subscription for feature-c: %v", err)
+	}
+	// --------------------------
+
+	// 2. Run the sync with two features redirecting to the same target
+	// feature-a -> feature-c
+	// feature-b -> feature-c
+	desiredState := []WebFeature{
+		{FeatureKey: "feature-c", Name: "Feature C", Description: "", DescriptionHTML: ""},
+	}
+	opts := []SyncWebFeaturesOption{
+		WithRedirectTargets(map[string]string{
+			"feature-a": "feature-c",
+			"feature-b": "feature-c",
+		}),
+	}
+
+	if err := spannerClient.SyncWebFeatures(ctx, desiredState, opts...); err != nil {
+		t.Fatalf("SyncWebFeatures failed: %v", err)
+	}
+
+	// 3. Verify final state
+	// feature-c should still have a saved search (it should be preserved as per logic 'Target already has one')
+	smsCFinal, err := spannerClient.GetSystemManagedSavedSearchByFeatureID(ctx, featureKeyToID["feature-c"])
+	if err != nil {
+		t.Fatalf("Failed to get system managed saved search for feature-c after sync: %v", err)
+	}
+
+	// It should be the original one because the target already existed.
+	if smsCFinal.SavedSearchID != smsC.SavedSearchID {
+		t.Errorf("Expected feature-c saved search ID to be preserved. Got %s, want %s",
+			smsCFinal.SavedSearchID, smsC.SavedSearchID)
+	}
+
+	// --- Verify Subscription Deduplication ---
+	// Subscription A should be deleted (cascaded from SavedSearch A deletion).
+	// It was dropped during migration because user already has C.
+	_, err = spannerClient.GetSavedSearchSubscription(ctx, *subA, userID)
+	if !errors.Is(err, ErrMissingRequiredRole) {
+		// We expect a permission error here because the subscription was deleted, but if we get a different error,
+		// that's unexpected.
+		t.Errorf("Expected Subscription A to be deleted immediately, got error: %v", err)
+	}
+
+	// Subscription C should still exist.
+	_, err = spannerClient.GetSavedSearchSubscription(ctx, *subC, userID)
+	if err != nil {
+		t.Errorf("Expected subscription C to persist, got error: %v", err)
+	}
+	// -------------------------------------
+
+	// feature-a and feature-b saved searches should be deleted (cascaded from SystemManagedSavedSearches)
+	_, err = spannerClient.GetSystemManagedSavedSearchByFeatureID(ctx, featureKeyToID["feature-a"])
+	if !errors.Is(err, ErrQueryReturnedNoResults) {
+		t.Errorf("Expected feature-a system managed saved search to be deleted, got error: %v", err)
+	}
+	_, err = spannerClient.GetSystemManagedSavedSearchByFeatureID(ctx, featureKeyToID["feature-b"])
+	if !errors.Is(err, ErrQueryReturnedNoResults) {
+		t.Errorf("Expected feature-b system managed saved search to be deleted, got error: %v", err)
+	}
+
+	// SavedSearch A and B should be deleted (cascaded from GetChildDeleteKeyMutations)
+	_, err = spannerClient.GetSavedSearch(ctx, smsA.SavedSearchID)
+	if !errors.Is(err, ErrQueryReturnedNoResults) {
+		t.Errorf("Expected SavedSearch for A to be deleted immediately, got error: %v", err)
+	}
+	_, err = spannerClient.GetSavedSearch(ctx, smsB.SavedSearchID)
+	if !errors.Is(err, ErrQueryReturnedNoResults) {
+		t.Errorf("Expected SavedSearch for B to be deleted immediately, got error: %v", err)
+	}
+
+	// Verify orphans are gone
+	_, err = spannerClient.GetSavedSearch(ctx, smsA.SavedSearchID)
+	if !errors.Is(err, ErrQueryReturnedNoResults) {
+		t.Errorf("Expected SavedSearch for A to be cleaned up, got error: %v", err)
+	}
+}

